### ISSUE TITLE
Issue with transfer learning of ASR model

### ISSUE THREAD

**Issue by gsoul** [NVIDIA/OpenSeq2Seq#256]
Hello, everyone.

First of all I'd like to thank authors for such an amazing repo and all the work that was put into it.

Currently I'm trying to use transfer learning of ASR model to further use it for 1 speaker. In order to achieve this I:
1. cloned this repo
2. installed the dependencies
3. prepared new data in the libri-like format
4. Split data into train/test portions
5. Downloaded the model and put it into "checkpoint" folder
6. modified w2lplus_large_8gous_mp.py file to use the new data, and decreased num_gpus to 2.

Command I use to run training:
`CUDA_VISIBLE_DEVICES=0,1 python3 run.py --config_file=example_configs/speech2text/new_data_test.py --mode=train_eval --continue_learning --logdir=checkpoint`

And the output I get:
```
task:0/device:GPU:0 with 10415 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080 Ti, pci bus id: 0000:4b:00.0, compute capability: 6.1)
WARNING:tensorflow:Error encountered when serializing REGULARIZATION_FUNCTIONS.
Type is unsupported, or the types of the items don't match field type in CollectionDef.
'tuple' object has no attribute 'name'
*** Running evaluation on a validation set:
***     Validation loss: 29.6603
***     Validation WER:  0.2448
WARNING:tensorflow:Error encountered when serializing REGULARIZATION_FUNCTIONS.
Type is unsupported, or the types of the items don't match field type in CollectionDef.
'tuple' object has no attribute 'name'
WARNING:tensorflow:Error encountered when serializing REGULARIZATION_FUNCTIONS.
Type is unsupported, or the types of the items don't match field type in CollectionDef.
'tuple' object has no attribute 'name'
*** Epoch 3660, global step 109800: ***     Train loss: 244.9925
time per step = 0:00:5.564
***     Sample WER: 1.0000
***     Sample target:     our analysis of miss lucy ward's lodgings yielded two primary links in our long chain that leads us from victim to culprit
***     Sample prediction:
*** Epoch 3660, global step 109810: ***     Train loss: 219.4841
time per step = 0:00:10.105
*** Running evaluation on a validation set:
2018-10-26 23:48:00.386912: E tensorflow/core/kernels/data/cache_dataset_ops.cc:314] The calling iterator did not fully read the dataset we were attempting to cache. In order to avoid unexpected truncation of the sequence, the current [partially cached] sequence will be dropped. This can occur if you have a sequence similar to `dataset.cache().take(k).repeat()`. Instead, swap the order (i.e. `dataset.take(k).cache().repeat()`)
***     Validation loss: 239.8677
***     Validation WER:  1.0000
```

2 issues I have with it:
1. It seems that network applies trained coefficients only at the very beginning before 1st step of training. Starting with 1st step the error jumps to 200+ values, which is too much.
2. For some reason it doesn't use 2nd GPU.

Here you can find the full output I get: https://pastebin.com/ap0PRdrz

Please advise what I might be doing wrong to do proper transfer learning.

**Comment by vsl9** [comment]
Hello Konstantin,

Thanks for your interest in OpenSeq2Seq.

> It seems that network applies trained coefficients only at the very beginning before 1st step of training. Starting with 1st step the error jumps to 200+ values, which is too much.

We'll try to look into that next week. 

> For some reason it doesn't use 2nd GPU.

Do you use Horovod? `w2lplus_large_8gpus_mp.py` has enabled Horovod mode by default: https://github.com/NVIDIA/OpenSeq2Seq/blob/87e9625af99b799808d5d9af8147f8ee4a2c5dbe/example_configs/speech2text/w2lplus_large_8gpus_mp.py#L15

To start train with Horovod, one need to use mpiexec/mpirun:
```
mpiexec --allow-run-as-root -np NUMBER_OF_GPUs python run.py ...OTHER_ARGS...
```
For 2 GPUs it might be better just to disable Horovod either in the config file, or by adding  `--use_horovod=False` argument to the command.

**Comment by gsoul** [comment]
Yes, I used Horovod. And thank you, I'll try to disable it.

I'm looking forward to hear about the transfer learning issue.

If that's important I use:
- Ubuntu 16.04.1 LTS
- Python 3.6.0 :: Anaconda 4.3.0 (64-bit)
- tensorflow-gpu==1.8.0

and 2 * 1080ti GPUs

And again, thank you guys.

**Comment by gsoul** [comment]
Hey, guys.

Sorry, but any updates here?

**Comment by blisc** [comment]
Can you confirm that you are using a mixed precision model on an 1080 ti? 1080 ti does not support float16.

**Comment by gsoul** [comment]
1. Yes, I use 1080ti to fine-tune "w2lplus_large_8gous_mp" model.
2. Lower speed is not an issue for me right now
3. I remember being able to train nv-wavenet in fp16 mode on my gpus (perhaps fp16 thing was emulated, but allowed to increase batch size ~2x)

The issue for me is that for some reason it doesn't continue from checkpoint, but starts from scratch. And while I think my hardware setup would be enough for fine-tuning, I have doubts about training from the very beginning.

**Comment by blisc** [comment]
I have identified two issues that could be preventing you from fine tuning the model on a new dataset.
1) 1080 ti does not support float16 and by extension does not support mixed precision. If you run use either mode, you are running your model on your CPU.
2) The saved checkpoint for the w2lplus_large_8gpus_mp has a learning rate that decays to 0, which makes fine tuning difficult.

I would suggest to try using the [Wave2Letter+-34](https://drive.google.com/file/d/1hI9Rv_px5vqpuWQOCwfKmZzRVXMPiTtT/view) with the [w2lplus_large_8gpus_float_test](https://github.com/NVIDIA/OpenSeq2Seq/blob/transfer_learning_dtype/example_configs/speech2text/w2lplus_xlarge_34_8gpus_float_test.py) config under the [transfer_learning_dtype](https://github.com/NVIDIA/OpenSeq2Seq/tree/transfer_learning_dtype/) branch.

The new branch has an experimental feature where we can load mixed precision models into float32 so that we can solve issue 1. The config file has also been modified to use a fixed learning rate to solve issue 2, although we can only use Wave2Letter+-34 and Wave2Letter+-54-syn since the base Wave2Letter+ model has the learning rate decay difficulty.

Let me know if this helps.

**Comment by blisc** [comment]
Great to hear that it is working. Yes it does a larger amount of time to start due to the mixed -> float32 conversion at the beginning of training.
